### PR TITLE
feat: android: setup config with initialization

### PR DESCRIPTION
**Pull request by mrehan27**
### Tickets

- Closes: [7657](https://github.com/customerio/issues/issues/7657) - [Android] Initialise react native package
- Closes: [7668](https://github.com/customerio/issues/issues/7668) - [Android] Rich push notifications capability
- Closes: [7918](https://github.com/customerio/issues/issues/7918) - [Android] Support for In-app messaging

### Changes

- Added native classes to parse and config values from react and pass to Android SDK

### Pending Items

- Update Android SDK to support more config options to align with iOS SDK

### CODE REVIEW

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt [hunk 1]
```diff
@@ -27,7 +27,7 @@ class CustomerIOReactNativeModule(
         val isDebuggable = 0 != reactContext.applicationInfo.flags and
                 ApplicationInfo.FLAG_DEBUGGABLE
         CustomerIOReactNativeInstance.setLogLevel(
-            logLevel = if (isDebuggable) CioLogLevel.DEBUG else CioLogLevel.NONE,
+            logLevel = if (isDebuggable) CioLogLevel.DEBUG else CioLogLevel.ERROR,
```
Comment (ami-aman) [inline android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:30]: There are 4 types of log levels - info, none, error & debug. This condition satisfies only two of them.

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt [hunk 2]
```diff
@@ -1,105 +1,177 @@
...
+                ApplicationInfo.FLAG_DEBUGGABLE
+        CustomerIOReactNativeInstance.setLogLevel(
+            logLevel = if (isDebuggable) CioLogLevel.DEBUG else CioLogLevel.ERROR,
+        )
     }
 
-    fun logMessage(message: String) {
-        Log.d("[CIO]", message)
+    override fun getName(): String {
+        return MODULE_NAME
```
Comment (ami-aman) [inline android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:35]: What is the expected value of `MODULE_NAME` ?

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt [hunk 3]
```diff
@@ -1,105 +1,177 @@
...
+            )?.takeIfNotBlank().toRegion()
+            val organizationId = env.getProperty<String>(
+                Keys.Environment.ORGANIZATION_ID
+            )?.takeIfNotBlank()
+
+            customerIO = CustomerIO.Builder(
+                siteId = siteId,
+                apiKey = apiKey,
+                region = region,
+                appContext = reactApplicationContext.applicationContext as Application,
```
Comment (ami-aman) [inline android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:79]: Just curious to understand what `appContext` does ?

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt [hunk 4]
```diff
@@ -1,105 +1,177 @@
...
-    fun config(attributes: ReadableMap?) {
-        if (isInitialized()) {
-            logMessage("Customer.io instance already initialized")
-            return
+    private fun CustomerIO.Builder.setupConfig(config: Map<String, Any>?): CustomerIO.Builder {
+        if (config == null) return this
+
+        val logLevel = config.getProperty<Double>(Keys.Config.LOG_LEVEL).toCIOLogLevel()
+        CustomerIOReactNativeInstance.setLogLevel(logLevel = logLevel)
+        setLogLevel(level = logLevel)
```
Comment (ami-aman) [inline android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:104]: This is confusing me, you are setting `logLevel` using `CustomerIOReactNativeInstance` and somewhere in this class, how are these two different ?

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt (hunk 1 above)
Comment (mrehan27) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:30]: Yes, we do have 4. These log types are handled in `TypeConversion.kt`. This method just sets log based on build type until SDK is initialized. As soon as the SDK is initialized, `logLevel` is updated to match user preference 🤓

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt (hunk 3 above)
Comment (mrehan27) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:79]: `appContext` is `Context` instance which is the core class of Android. It lets you assess resources, identify build types, start screens, and anything that is directly linked with Android OS 🤓

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt [hunk 5]
```diff
@@ -1,105 +1,177 @@
...
+            )?.takeIfNotBlank()
+
+            customerIO = CustomerIO.Builder(
+                siteId = siteId,
+                apiKey = apiKey,
+                region = region,
+                appContext = reactApplicationContext.applicationContext as Application,
+            ).apply {
+                // TODO: Update when sdk version is released
+                setClient(Client.ReactNative)
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:82]: Yeah, good suggestion. I'm keeping [7672](https://github.com/customerio/issues/issues/7672) on board till we fix this.

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt (hunk 4 above)
Comment (mrehan27) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:104]: The first `logLevel` is default value based on build type as we don't have user preference at that point. As soon as we get the user preference value, we update if for both `CustomerIOReactNativeInstance` (ReactNative SDK) and `setLogLevel` (Android SDK)

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ami-aman) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:82]: 👍

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt (hunk 4 above)
Comment (ami-aman) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:104]: By `CustomerIOReactNativeInstance` you mean the android side of react package for which you have created logger class to show logs based on preference? Please correct me if I am wrong.

**Review by ami-aman — APPROVED** [review]
👍

File: android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt (hunk 4 above)
Comment (mrehan27) [reply android/src/main/java/io/customer/reactnative/sdk/CustomerIOReactNativeModule.kt:104]: Yeah exactly, you are correct. `CustomerIOReactNativeInstance` is just holding stuff for Android part of react package to make communication with our Android SDK easier.

**Pull request merged**